topiary-cli/src/configuration.rs

```diff
@@ -1,73 +1,218 @@
+                } else {
+                    log::warn!(
+                        "Could not find configuration file: {}",
+                        candidate.to_string_lossy()
+                    );
+
+                    ConfigSource::Missing
+                }
+            }
+        }
     }
+}
+
+impl TryFrom<&ConfigSource> for toml::Value {
+    type Error = TopiaryError;
 
-    // Otherwise consider the configuration files in order. Lowest priority first:
-    //   - The built-in configuration `default_configuration_toml`
-    //   - `~/.config/topiary/languages.toml` (or equivalent)
-    //   - `.topiary/languages.toml`
-    //   - `config_file` as passed by `--configuration_file/-c/TOPIARY_CONFIGURATION_FILE`
+    fn try_from(source: &ConfigSource) -> Result<Self, Self::Error> {
+        match source {
+            ConfigSource::Builtin => Ok(default_configuration_toml()),
+
+            ConfigSource::File(file) => {
+                let config = std::fs::read_to_string(file)?;
+                toml::from_str(&config).map_err(TopiaryError::from)
+            }
+
+            ConfigSource::Missing => Err(TopiaryError::Bin(
+                "Could not parse missing configuration".into(),
+                None,
+            )),
+        }
+    }
+}
+
+/// Return the valid sources of configuration, in priority order (lowest to highest):
+///
+/// 1. Built-in configuration (`topiary::default_configuration_toml`)
+/// 2. `~/.config/topiary/languages.toml` (or equivalent)
+/// 3. `.topiary/languages.toml` (or equivalent)
+/// 4. `file`, passed as a CLI argument/environment variable
+fn configuration_sources(file: &Option<PathBuf>) -> Vec<ConfigSource> {
     [
-        Some(find_configuration_dir()),
-        find_workspace_configuration_dir(),
-        config_file,
+        ConfigSource::Builtin,
+        Some(find_os_configuration_dir()).into(),
+        find_workspace_configuration_dir().into(),
+        file.clone().into(),
     ]
     .into_iter()
-    .filter_map(|path| {
-        path.map(|p| match p.is_file() {
-            // The path already points to a file, assume the file is the configuration file
-            true => p,
-            // The path points to a directory, assume it is a topiary configuration directory and append "languages.toml"
-            false => p.join("languages.toml"),
-        })
-    })
-    .filter_map(|file| -> Option<Result<toml::Value, toml::de::Error>> {
-        std::fs::read_to_string(file)
-            .map(|config| toml::from_str(&config))
-            .ok()
-    })
-    .try_fold(default_configuration_toml(), |a, b| {
-        let b = b?;
-        Ok(merge_toml_values(a, b, 3))
-    })
+    .filter(ConfigSource::is_valid)
+    .collect()
+}
+
+/// Consume configuration and collate as specified
+fn configuration_toml(
+    sources: &[ConfigSource],
+    collation: &CollationMode,
+) -> CLIResult<toml::Value> {
+    match collation {
+        CollationMode::Merge => todo!(),
```

Comment:
I think it might be possible for merge to be `merge_toml_values` with a higher `merge_depth`.

Reply:
Yep: That's what I did in #596 :+1: